arangod/Aql/RemoteExecutor.cpp

```diff
@@ -0,0 +1,483 @@
+            VPackSlice code = it.get("code");
+            VPackSlice message = it.get("message");
+            if (code.isNumber() && message.isString()) {
+              query->registerWarning(code.getNumericValue<int>(),
+                                     message.copyString().c_str());
+            }
+          }
+        }
+      }
+      if (slice.hasKey("code")) {
+        return {ExecutionState::DONE, slice.get("code").getNumericValue<int>()};
+      }
+    }
+
+    return {ExecutionState::DONE, TRI_ERROR_INTERNAL};
+  }
+
+  // For every call we simply forward via HTTP
+  VPackBuilder bodyBuilder;
+  bodyBuilder.openObject();
+  bodyBuilder.add("code", VPackValue(errorCode));
+  bodyBuilder.close();
+
+  auto bodyString = std::make_shared<std::string const>(bodyBuilder.slice().toJson());
+
+  auto res = sendAsyncRequest(rest::RequestType::PUT, "/_api/aql/shutdown/", bodyString);
+  if (!res.ok()) {
+    THROW_ARANGO_EXCEPTION(res);
+  }
+  return {ExecutionState::WAITING, TRI_ERROR_NO_ERROR};
+}
+
+Result ExecutionBlockImpl<RemoteExecutor>::sendAsyncRequest(
+    arangodb::rest::RequestType type, std::string const& urlPart,
+    std::shared_ptr<std::string const> body) {
+  auto cc = ClusterComm::instance();
+  if (cc == nullptr) {
+    // nullptr only happens on controlled shutdown
+    return {TRI_ERROR_SHUTTING_DOWN};
+  }
+
+  // Later, we probably want to set these sensibly:
+  CoordTransactionID const coordTransactionId = TRI_NewTickServer();
+  std::unordered_map<std::string, std::string> headers;
+  if (!_ownName.empty()) {
+    headers.emplace("Shard-Id", _ownName);
+  }
+
+  std::string url = std::string("/_db/") +
+                    arangodb::basics::StringUtils::urlEncode(
+                        _engine->getQuery()->trx()->vocbase().name()) +
+                    urlPart + _queryId;
+
+  ++_engine->_stats.requests;
+  std::shared_ptr<ClusterCommCallback> callback =
+      std::make_shared<WakeupQueryCallback>(this, _engine->getQuery());
+
+  // TODO Returns OperationID do we need it in any way?
+  _lastTicketId =
+      cc->asyncRequest(coordTransactionId, _server, type, url, std::move(body),
+                       headers, callback, defaultTimeOut, true);
+
+  return {TRI_ERROR_NO_ERROR};
+}
+
+bool ExecutionBlockImpl<RemoteExecutor>::handleAsyncResult(ClusterCommResult* result) {
+  if (_lastTicketId == result->operationID) {
+    // TODO Handle exceptions thrown while we are in this code
+    // Query will not be woken up again.
+    _lastError = handleCommErrors(result);
+    if (_lastError.ok()) {
+      _lastResponse = result->result;
+    }
+    _lastTicketId = 0;
+  }
+  return true;
+}
+
+arangodb::Result ExecutionBlockImpl<RemoteExecutor>::handleCommErrors(ClusterCommResult* res) const {
+  if (res->status == CL_COMM_TIMEOUT || res->status == CL_COMM_BACKEND_UNAVAILABLE) {
+    return {res->getErrorCode(), res->stringifyErrorMessage()};
+  }
+  if (res->status == CL_COMM_ERROR) {
```

Comment:
Just so it is not forgotten, I think we should verify this error handling code and align it with the code in ClusterMethods.cpp.

It seems that depening on the codepath used this check might be incorrect, I think this is only ever set if `res->result == nullptr`. At least that seems to be the case in https://github.com/arangodb/arangodb/blob/devel/arangod/Cluster/ClusterComm.h#L247

Reply:
I opened an internal ticket so this won't be forgotten.